### ISSUE TITLE
Facing echo issue

### ISSUE THREAD

**Issue by kairavi-magnates** [juha-h/baresip-studio#122]
While calling using baresip application i am facing echo issue in audio call. I have already checked Acoustic Echo Cancellation in Audio setting.

Can you please share your inputs?

**Comment by juha-h** [comment]
kairavi-magnates writes:

> While calling using baresip application i am facing echo issue in
> audio call. I have already checked Acoustic Echo Cancellation in Audio
> setting.

I can only suggest to check that speakerphone is not activated and that
speaker volume is not too high.

I myself have several devices and don't have echo problems with them.

**Comment by kairavi-magnates** [comment]
Thanks for the quick response. 

The speaker is already off. I have just observe that issue is with android version 9 device (one plus 3).

**Comment by juha-h** [comment]
kairavi-magnates writes:

> The speaker is already off. I have just observe that issue is with
> android version 9 device (one plus 3).

Must be One Plus related, since I don't experience echo on my Nokia 3.1
running Android 9.

**Comment by kairavi-magnates** [comment]
Apart from the possibility of device os version, can we do something with audio codec configuration?

**Comment by juha-h** [comment]
I don't think that choice of codec matters, but you can try if there is any difference between codecs with 8000 and 16000 rates, eg. PCMU and opus.

**Comment by juha-h** [comment]
I forgot to mention that there is Default Call Volume setting that you can experiment with.  It sets the volume to the given value when call is established. I usually set it to value 6 or 7, which gives me comfortable volume without echo.

**Comment by juha-h** [comment]
I'll close this now, since there is nothing else I can do to help.

**Comment by juha-h** [comment]
@kairavi-magnates See https://github.com/juha-h/baresip-studio/issues/188#issuecomment-2629854735.

**Comment by juha-h** [comment]
@kairavi-magnates @rabongithub @morty303 @goderfly @adatum @kar-l  Latest version of baresip available from F-Droid (64.0.0)  is now able to use hardware Echo Canceler of the device.

**Comment by morty303** [comment]
does redmi note 10 pro have this?
Hardware Echo Canceler
do i need to enable this?

**Comment by juha-h** [comment]
You don't need to enable hw AEC.  You can check from logcat if AEC is supported.  You will get this to logcat at start if hw AEC is not available:
```
Log.i(TAG, "Hardware AEC is NOT available")
```
Then when call starts, one of these appear in logcat:
```
                 if (aecAvailable) {
                    aec = AcousticEchoCanceler.create(recorderSessionId)
                    if (aec != null) {
                        if (!aec!!.getEnabled()) {
                            aec!!.setEnabled(true)
                            if (aec!!.getEnabled())
                                Log.d(TAG, "AEC is enabled")
                            else
                                Log.w(TAG, "Failed to enable AEC")
                        }
                        else
                            Log.d(TAG, "AEC is already enabled")
                    } else
                        Log.w(TAG, "Failed to create AEC for session $recorderSessionId")
                }
```

**Comment by morty303** [comment]
> You don't need to enable hw AEC. You can check from logcat if AEC is supported. You will get this to logcat at start if hw AEC is not available:
> 
> ```
> Log.i(TAG, "Hardware AEC is NOT available")
> ```
> 
> Then when call starts, one of these appear in logcat:
> 
> ```
>                  if (aecAvailable) {
>                     aec = AcousticEchoCanceler.create(recorderSessionId)
>                     if (aec != null) {
>                         if (!aec!!.getEnabled()) {
>                             aec!!.setEnabled(true)
>                             if (aec!!.getEnabled())
>                                 Log.d(TAG, "AEC is enabled")
>                             else
>                                 Log.w(TAG, "Failed to enable AEC")
>                         }
>                         else
>                             Log.d(TAG, "AEC is already enabled")
>                     } else
>                         Log.w(TAG, "Failed to create AEC for session $recorderSessionId")
>                 }
> ```

thanks bro. nice from you.
i am no dev but i will check this later.
